vir/src/callable_idents.rs

```diff
@@ -1,36 +1,44 @@
 use crate::{ExprGen, PredicateAppGen, Type, PredicateAppGenData, StmtGenData, MethodCallGenData, VirCtxt, TypeData, DomainParamData};
 use sealed::sealed;
 
-pub trait CallableIdent<'vir, A: Arity<'vir>> {
-    fn new(name: &'vir str, args: A) -> Self;
+pub trait CallableIdent<'vir, A: Arity<'vir>, ResultTy> {
+    fn new(name: &'vir str, args: A, result_ty: ResultTy) -> Self;
     fn name(&self) -> &'vir str;
     fn arity(&self) -> &A;
+    fn result_ty(&self) -> ResultTy;
 }
-pub trait ToKnownArity<'vir, T: 'vir>: CallableIdent<'vir, UnknownArityAny<'vir, T>> + Sized {
-    fn to_known<'tcx, K: CallableIdent<'vir, KnownArityAny<'vir, T, N>>, const N: usize>(self) -> K {
-        K::new(self.name(), KnownArityAny::new(self.arity().args().try_into().unwrap()))
+pub trait ToKnownArity<'vir, T: 'vir, ResultTy>: CallableIdent<'vir, UnknownArityAny<'vir, T>, ResultTy> + Sized {
+    fn to_known<'tcx, K: CallableIdent<'vir, KnownArityAny<'vir, T, N>, ResultTy>, const N: usize>(self) -> K {
+        K::new(
+            self.name(),
+            KnownArityAny::new(self.arity().args().try_into().unwrap()),
+            self.result_ty(),
+        )
     }
 }
-impl<'vir, T: 'vir, K: CallableIdent<'vir, UnknownArityAny<'vir, T>>> ToKnownArity<'vir, T> for K {}
+impl<'vir, T: 'vir, ResultTy, K: CallableIdent<'vir, UnknownArityAny<'vir, T>, ResultTy>> ToKnownArity<'vir, T, ResultTy> for K {}
 
 #[derive(Debug, Clone, Copy)]
-pub struct FunctionIdent<'vir, A: Arity<'vir>>(&'vir str, A);
-impl<'vir, A: Arity<'vir>> CallableIdent<'vir, A> for FunctionIdent<'vir, A> {
-    fn new(name: &'vir str, args: A) -> Self {
-        Self(name, args)
+pub struct FunctionIdent<'vir, A: Arity<'vir>>(&'vir str, A, Type<'vir>);
+impl<'vir, A: Arity<'vir>> CallableIdent<'vir, A, Type<'vir>> for FunctionIdent<'vir, A> {
+    fn new(name: &'vir str, args: A, result_ty: Type<'vir>) -> Self {
+        Self(name, args, result_ty)
     }
     fn name(&self) -> &'vir str {
         self.0
     }
     fn arity(&self) -> &A {
         &self.1
     }
+    fn result_ty(&self) -> Type<'vir> {
+        self.2
+    }
 }
 
 #[derive(Debug, Clone, Copy)]
 pub struct MethodIdent<'vir, A: Arity<'vir>>(&'vir str, A);
-impl<'vir, A: Arity<'vir>> CallableIdent<'vir, A> for MethodIdent<'vir, A> {
-    fn new(name: &'vir str, args: A) -> Self {
+impl<'vir, A: Arity<'vir>> CallableIdent<'vir, A, ()> for MethodIdent<'vir, A> {
+    fn new(name: &'vir str, args: A, _unused: ()) -> Self {
```

Comment:
Will we try to use this mechanism also to check method returns? In general in Viper, methods can return any number of values, which sounds like `Arity` yet again.